### PR TITLE
feat(auth-ui): Create Country search dialog and reusable phone text field

### PR DESCRIPTION
**Pull request by wizlif**
- Reusable Phone input.
- Country search dialog.

Signed-off-by: Isaac <wizlif.144@gmail.com>

### CODE REVIEW

File: lib/presentation/authentication/auth_screen.dart [hunk 1]
```diff
@@ -128,104 +111,27 @@ class _AuthPageState extends State<AuthPage> {
...
-                                                      .value.text.isEmpty
-                                                  ? kInactiveColor
-                                                  : kErrorColor),
-                                    ),
-                                  ),
-                                ),
-                              ],
-                            ),
+                            PhoneInput(isValid: (valid) {
+                              setState(() {
```
Comment (Xazin) [inline lib/presentation/authentication/auth_screen.dart]: Use short-hand syntax `=>`

File: lib/presentation/shared_widgets/country_search_dialog.dart [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    return SizedBox(
+      width: double.maxFinite,
+      child: Column(
+        mainAxisSize: MainAxisSize.min,
+        children: [
+          TextField(
+            decoration: const InputDecoration(
+              prefixIcon: Icon(Icons.search),
+            ),
+            onChanged: (text) {
```
Comment (Xazin) [inline lib/presentation/shared_widgets/country_search_dialog.dart]: Can we abstract this to a method to keep the UI code a bit cleaner

File: lib/presentation/shared_widgets/country_search_dialog.dart [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+              _countries.clear();
+
+              if (text.isEmpty) {
+                setState(() {
+                  _countries.addAll(countries);
+                });
+              } else {
+                setState(() {
```
Comment (Xazin) [inline lib/presentation/shared_widgets/country_search_dialog.dart]: Use short-hand syntax `=>`

File: lib/presentation/shared_widgets/country_search_dialog.dart [hunk 4]
```diff
@@ -0,0 +1,88 @@
...
+        children: [
+          TextField(
+            decoration: const InputDecoration(
+              prefixIcon: Icon(Icons.search),
+            ),
+            onChanged: (text) {
+              _countries.clear();
+
+              if (text.isEmpty) {
+                setState(() {
```
Comment (Xazin) [inline lib/presentation/shared_widgets/country_search_dialog.dart]: Use short-hand syntax `=>`

File: lib/presentation/shared_widgets/country_search_dialog.dart [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+            },
+          ),
+          Expanded(
+            child: ListView.builder(
+              shrinkWrap: true,
+              itemBuilder: (context, index) {
+                final country = _countries[index];
+
+                return InkWell(
+                  onTap:(){
```
Comment (Xazin) [inline lib/presentation/shared_widgets/country_search_dialog.dart]: add a space between () {

File: lib/presentation/shared_widgets/country_search_dialog.dart [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+                    if(widget.onCountrySelected != null){
```
Comment (Xazin) [inline lib/presentation/shared_widgets/country_search_dialog.dart]: add a space between if( and ){

File: lib/presentation/shared_widgets/phone_input.dart [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+  bool validatedNumber = false;
+
+  @override
+  void initState() {
+    if (widget.phone != null) {
+      final country =
+          countries.where((country) => country.code == widget.phone?.country);
+
+      /// Strip country
+      if (country.isNotEmpty) {
```
Comment (Xazin) [inline lib/presentation/shared_widgets/phone_input.dart]: Maybe we can shorten this a bit like this
```
_selected = country.isNotEmpty ? country.first : countries.where((country) => country.code == "NL").first;
```

File: lib/presentation/shared_widgets/phone_input.dart [hunk 8]
```diff
@@ -0,0 +1,190 @@
...
+    }
+
+    _phoneNumberController.addListener(() {
+      if (_phoneNumberController.value.text.isNotEmpty) {
+        setState(() {
+          _validatePhone(_selected!, _phoneNumberController.value.text);
+        });
+      }
+    });
+    super.initState();
```
Comment (Xazin) [inline lib/presentation/shared_widgets/phone_input.dart]: We should move super.initState(); to the first line of the initState method, it should always be called first.

File: lib/presentation/shared_widgets/phone_input.dart [hunk 9]
```diff
@@ -0,0 +1,190 @@
...
+                      ),
+                      const Icon(Icons.expand_more_outlined),
+                    ],
+                  ),
+                ),
+              ),
+              const SizedBox(width: 10),
+              Expanded(
+                child: TextFormField(
+                  controller: _phoneNumberController,
```
Comment (Xazin) [inline lib/presentation/shared_widgets/phone_input.dart:97]: Can we format a phone number with the correct format (region-based?) with a listener on this controller or in the onChange method? Do you want to look into this or should I? :D

File: lib/presentation/shared_widgets/phone_input.dart [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+        )
+      ],
+    );
+  }
+
+  void _showCountrySelectDialog() {
+    showDialog<void>(
+      context: context,
+      barrierDismissible: true, // user must tap button!
+      builder: (BuildContext context) {
```
Comment (Xazin) [inline lib/presentation/shared_widgets/phone_input.dart]: We can also use short-hand syntax for this `=>`

File: test/widgets/phone_input_test.dart [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+
+    });
+
+
+    testWidgets('Invalid Phone number triggers callback with false',
+        (WidgetTester tester) async {
+
+      final phoneInput = PhoneInput(
+          phone: PhoneResponse("NL","31 77891635"),
+        isValid:(valid){
```
Comment (Xazin) [inline test/widgets/phone_input_test.dart]: Add a space between ) {

**Review by Xazin — COMMENTED** [review]
Left a few comments, looking good though!

File: lib/presentation/shared_widgets/country_search_dialog.dart (hunk 2 above)
Comment (wizlif) [reply lib/presentation/shared_widgets/country_search_dialog.dart]: Sure let me move it to a `private` method

File: lib/presentation/shared_widgets/phone_input.dart (hunk 9 above)
Comment (wizlif) [reply lib/presentation/shared_widgets/phone_input.dart:97]: I had actually used the `PhoneResponse` class in `onChange` callback that provides the county code and formatted phone number (with region) whenever anything changes. Is that what you meant ??

**Pull request merged**